### PR TITLE
Issue 115 - Support return preference

### PR DESCRIPTION
**Pull request by lmsurpre**
### issue #115 - refactor FHIRPersistence interface to return Response wrappers. 

now the PersistenceLayer can return OperationOutcome AND Resource(s) at the same time which will be needed for honoring the return preference indicated by the client.

###  issue #115 - use return preference from Prefer header

1. add support for setting a default HTTPReturnPreference in fhir-client
2. introduce create and update tests for checking the response to returnPreferences
3. add OperationOutcome to FHIRRestOperationResponse
4. set FHIRRequestContext.HttpReturnPreference in the FHIRRestServletFilter (default="minimal")
5. update return entity based on HTTPReturnPreference in FHIRRequestContext
6. update BundleTest to pass return=representation since the tests depend on this

### CODE REVIEW

**Comment by prb112** [comment]
Does not appear upload logs happened on this PR.  I kicked off another build just in case.

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java [hunk 1]
```diff
@@ -219,6 +222,39 @@ public FHIRResponse conditionalCreate(JsonObject resource, FHIRParameters parame
...
+        int headersSize = (headers != null ? headers.length : 0);
+        FHIRRequestHeader[] result = new FHIRRequestHeader[headersSize + 1];
+        if (headers != null) {
+            for (int i = 0; i < headers.length; i++) {
+                result[i] = headers[i];
+            }
+        }
+        
+        // Add the new header at the end.
+        result[result.length - 1] = preferHeader;
```
Comment (prb112) [inline fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: What happens if result.length = 0?

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 2]
```diff
@@ -653,7 +687,37 @@ protected void storeSearchParameters(Resource fhirResource, com.ibm.fhir.persist
      * @throws JAXBException
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java]: strange are we actually throwing this?

File: fhir-persistence-jdbc/src/test/java/testng.xml [hunk 3]
```diff
@@ -1,3 +1,4 @@
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
```
Comment (prb112) [inline fhir-persistence-jdbc/src/test/java/testng.xml]: 👍 

annoys me on every build :)

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java]: I see this in a bunch of javadoc on FHIRPersistenceJDBCImpl, but i'm guess its just carry-over from "back in the day".  Let me remove it across the board quick...

File: fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java (hunk 1 above)
Comment (lmsurpre) [reply fhir-client/src/main/java/com/ibm/fhir/client/impl/FHIRClientImpl.java]: its initialized just above with `FHIRRequestHeader[] result = new FHIRRequestHeader[headersSize + 1];` ...I don't think length can be 0 here

**Comment by JohnTimm** [comment]
Could this be refactored to remove the SuppressWarnings?

```
        public Builder<T> resource(@SuppressWarnings("unchecked") T... resource) {
            for (T value : resource) {
                this.resource.add(value);
            }
            return this;
        }
```

File: fhir-persistence/src/main/java/com/ibm/fhir/persistence/MultiResourceResult.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    public boolean isSuccess() {
+        return success;
+    }
+    /**
+     * The resources returned from the interaction
+     * 
+     * @return
+     *     An unmodifiable list containing immutable objects of type {@link Resource}.
+     */
+    public List<T> getResource() {
```
Comment (prb112) [inline fhir-persistence/src/main/java/com/ibm/fhir/persistence/MultiResourceResult.java]: any reason not to put getResources vs Resource?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/BulkExportBatchLet.java [hunk 5]
```diff
@@ -301,7 +301,7 @@ public String process() throws Exception {
             FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
             txn.begin();
             persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, searchContext);
-            List<Resource> resources = fhirPersistence.search(persistenceContext, resourceType);
+            List<Resource> resources = fhirPersistence.search(persistenceContext, resourceType).getResource();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/BulkExportBatchLet.java]: I added a comment in multi response result.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/ChunkReader.java [hunk 6]
```diff
@@ -201,7 +201,7 @@ public Object readItem() throws Exception {
         FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
         txn.begin();
         persistenceContext = FHIRPersistenceContextFactory.createPersistenceContext(null, searchContext);
-        resources = fhirPersistence.search(persistenceContext, resourceType);
+        resources = fhirPersistence.search(persistenceContext, resourceType).getResource();
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/ChunkReader.java]: add a comment about getResource vs getResources in the multiple response.

File: fhir-model/src/main/java/com/ibm/fhir/model/util/FHIRUtil.java [hunk 7]
```diff
@@ -171,6 +173,13 @@
         UsageContext.class,
         Dosage.class));
     private static final Map<String, String> CONCRETE_TYPE_NAME_MAP = buildConcreteTypeNameMap();
+    private static final OperationOutcome ALL_OK = OperationOutcome.builder()
+                                                            .issue(Issue.builder()
+                                                                    .severity(IssueSeverity.INFORMATION)
```
Comment (prb112) [inline fhir-model/src/main/java/com/ibm/fhir/model/util/FHIRUtil.java]: is there an alternative class to put this ALL_OK. outcome in?

File: fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/spec/HistoryOperation.java [hunk 8]
```diff
@@ -31,7 +31,7 @@ public void process(TestContext tc) throws FHIRPersistenceException {
         
         final String logicalId = resource.getId().getValue();
         
-        List<Resource> resources = tc.getPersistence().history(context, resource.getClass(), logicalId);
+        List<? extends Resource> resources = tc.getPersistence().history(context, resource.getClass(), logicalId).getResource();
```
Comment (prb112) [inline fhir-persistence-jdbc/src/test/java/com/ibm/fhir/persistence/jdbc/test/spec/HistoryOperation.java]: semantic question getResources?

File: fhir-server-test/src/test/resources/test.properties [hunk 9]
```diff
@@ -32,4 +32,5 @@ test.kafka.topicName = fhirNotifications
 fhirclient.logging.enabled = false
 
 fhirclient.hostnameVerification.enabled = false
-fhirclient.http.receive.timeout = 60000
\ No newline at end of file
+fhirclient.http.receive.timeout = 60000
+fhirclient.http.return.pref = minimal
```
Comment (prb112) [inline fhir-server-test/src/test/resources/test.properties]: good thing to highlight that our tests by default return minimal

Comment (prb112) [reply fhir-server-test/src/test/resources/test.properties]: sorry Filter explains why...

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/ServerSpecTest.java [hunk 10]
```diff
@@ -81,6 +81,39 @@ public void testCreatePatient() throws Exception {
...
+        Patient responsePatient = response.readEntity(Patient.class);
+        assertNotNull(responsePatient);
+        
+        response = target.path("Patient").request()
+                .header("Prefer", "return=OperationOutcome")
+                .post(entity, Response.class);
+        assertResponse(response, Response.Status.CREATED.getStatusCode());
+        OperationOutcome oo = response.readEntity(OperationOutcome.class);
+        assertNotNull(oo);
+        assertTrue(oo.getIssue().toString().contains("dom-6: A resource should have narrative for robust management"));
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/ServerSpecTest.java]: Great Test! I didn't pick up this behavior from the code

File: fhir-persistence/src/main/java/com/ibm/fhir/persistence/MultiResourceResult.java (hunk 4 above)
Comment (prb112) [reply fhir-persistence/src/main/java/com/ibm/fhir/persistence/MultiResourceResult.java]: Net - follows the fhir-model use single vs plural 
done intentionally

File: fhir-model/src/main/java/com/ibm/fhir/model/util/FHIRUtil.java (hunk 7 above)
Comment (lmsurpre) [reply fhir-model/src/main/java/com/ibm/fhir/model/util/FHIRUtil.java]: FHIRUtil was the spot with all the buildOperationOutcome helpes, so it made the most sense.  If those ever get split out, this should get split out with them.

**Comment by prb112** [comment]
Fixed a bad merge

**Comment by prb112** [comment]
set DCO to pass.  The commit was a minor commit.

**Pull request merged**